File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.js
function solution(number, k) {
    /*
    1.아이디어
    자릿수와 상관없이 k개를 제거했을때 숫자의 자릿수는 모두 동일하므로
    가장 작은 숫자부터 제거해야한다.
    그러나.. 가장작은 숫자부터 제거했을때 반례가 존재함
    
    아직 k가 0보다 클때 담겨있는 값의 가장 뒷 값이 현재 값보다 작을 경우가 생기면 안된다.(최대가 아니므로)
    numbers배열을 돌면서 k>0 && result의 마지막 값이 현재값보다 작으면 pop() , k--
    조건을 만족하지 않을때까지 반복
    이후 현재값 push
    
    2. 시간복잡도
    n
    3. 자료구조
    결과값을 담을 stack

    */
    const result =[];
    
    for(let current of number){
        while(k>0 && result[result.length-1] < current){ 
            result.pop();
            k--;
        }
        result.push(current);
    }
    return result.splice(0,number.length-k).join('');
    
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;
import java.io.*;
/*
1. 아이디어
결과배열이 비어있으면 그냥 넣고,
원소가 존재하면 가장 끝에있는 원소와 현재 문자를 비교해서 현재원소가 더크면 결과배열에서 하나 꺼낸다.
더이상 만족하지 않을때 까지 반복한다. 이렇게 반복하면 결과배열은 큰숫자기준으로 정렬된다.

2. 시간복잡도
max:O(nlogn)

3. 자료구조
결과배열: 100만 = ArrayList
*/
class Solution {
    public String solution(String number, int k) {
        
        ArrayList<Integer> result= new ArrayList<>();
        
        for(char c : number.toCharArray()){
            int value= c- '0';
            while(k>0 && result.size()>0 && result.get(result.size()-1) < value){
                result.remove(result.size()-1);
                k--;
            }
            result.add(value);
        }  
        
       StringBuilder sb= new StringBuilder();
        
        for(int i=0;i<result.size()-k;i++){
            sb.append(result.get(i));
        }
        return sb.toString();

    }
}